### DISCUSSION TITLE
terminal

### DISCUSSION THREAD

**Discussion by blinkinglight** [coollabsio/coolify#3362] [General]
https://github.com/coollabsio/coolify/issues/2298

ssh can reuse connection

ssh config
    ControlMaster auto
    ControlPath ~/.ssh/sockets/%r@%h-%p
    ControlPersist 600
    
    or 
`ssh -t -o ControlPath=~/.ssh/master-someuniqname -o ControlMaster=auto -o ControlPersist=60 host cmd...`

**Comment by blinkinglight** [comment]
https://github.com/updroidinc/cmdr-pty?tab=readme-ov-file

**Reply by LEstradioto** [reply]
Go and Node supports long running process, like natively. PHP not, so we need to add dependencies or create one

**Reply by blinkinglight** [reply]
maybe then : https://reactphp.org/ ?

**Reply by LEstradioto** [reply]
yep, that would work to avoid node websocket, I added comments here where I do some suggestions, and reactphp is there:
https://github.com/coollabsio/coolify/pull/2586/files#diff-ad50ee06799e6d6c5ee7817d7e35a2f4d2820b6c1698d56ec9b7827d5e1ff161R37
still we need `node-pty` alternative too

**Reply by blinkinglight** [reply]
whats wrong with node-pty ?

**Reply by LEstradioto** [reply]
Oh, that issue discussion was going forward to wanting to remove node websocket and node-pty, to not use node at all, and not needing to install node as dependency somehow. Although Im in favor of utilizing note-pty as its well known, but since a xterm.js member commented there about creating node-pty from groundup on PHP, then... I think we should ponder about that.

**Reply by blinkinglight** [reply]
https://github.com/blinkinglight/cmdr-pty what do you think? cmd in readme. ( open in browser localhost:9990 )

**Reply by LEstradioto** [reply]
Man, but then you are using Go. Coolify is Laravel !

**Reply by blinkinglight** [reply]
coolify already uses go ;) or used but will use again - coollabsio/sentinel ;)

**Reply by LEstradioto** [reply]
sentinel is not built-in, probably will be an opt-in
if Andras want Terminal to be opt-in then using Node, Go, or whatever, is the same, since its an extra process in the end

**Comment by LEstradioto** [comment]
Here's the important part of ssh connection at Coolify [generateSshCommand](https://github.com/coollabsio/coolify/blob/c668d2d1ff1d51811fff1e9d54651546c0698c6a/bootstrap/helpers/remoteProcess.php#L133)

I see your suggestions being applied on `mux_enabled`. Im not sure how mux works of is it enabled by default?
I saw some interesting discussion here [about ssh and cpu usage spikes](https://github.com/coollabsio/coolify/issues/2110)
Since its high-priority issue, maybe this is being reworked.

On my PR #2586, the option changed is `RequestTTY=no` to then `RequestTTY=yes`. 

But when connection happens using SSH, websocket just re-uses the ptyProcess spawned by `node-pty`

**Reply by LEstradioto** [reply]
(so its just one ssh command at start in this case). I think to ssh being re-used across ALL Coolify requests, ssh must be reworked on Coolify itself.

**Reply by blinkinglight** [reply]
not sure if this is to the topic but found few days ago: https://github.com/pterodactyl with https://github.com/pterodactyl/wings and https://github.com/pterodactyl/panel

seems that they uses their own worker (wings golang) on servers and dashboard (php laravel)  maybe you could find some inspiration there? :)
